### PR TITLE
[multistage] decouple datablock from datatable

### PR DESCRIPTION
**Pull request by walterddr**
After #9561 . V2 engine no longer depend on anything datatable specific. It's time to decouple it

This PR is a single refactor that copied the DataTable API out into a DataBlock. So going forward any changes on datablock will not need to consider backward compatibility against data table. This opens up to use different serde encoding / remove not needed APIs later

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/datablock/DataBlock.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import java.math.BigDecimal;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.apache.pinot.common.response.ProcessingException;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.spi.utils.ByteArray;
+import org.roaringbitmap.RoaringBitmap;
+
+
+public interface DataBlock {
```
Comment (agavra) [inline pinot-common/src/main/java/org/apache/pinot/common/datablock/DataBlock.java:31]: I'm a bit confused, what was actually removed from `DataTable`?

Comment (walterddr) [reply pinot-common/src/main/java/org/apache/pinot/common/datablock/DataBlock.java:31]: This a single refactor that removed all datablock implement data table.

Comment (walterddr) [reply pinot-common/src/main/java/org/apache/pinot/common/datablock/DataBlock.java:31]: Next step is to remove all single access API and change serde method, or everything else

**Comment by siddharthteotia** [comment]
> So going forward any changes on datablock will not need to consider backward compatibility against data table.

I agree that we should try to decouple. But when leaf servers do the conversion of DataTable to Block (or whatever it is evolves to in future) to send to intermediary layer of servers, this logic should still continued to be written in compatible manner. Right ?

During deployment / rolling upgrade / restart, it is possible that intermediary servers and leaf servers are having different version.

File: pinot-common/src/main/java/org/apache/pinot/common/datatable/DataTableImplV4.java [hunk 2]
```diff
@@ -19,35 +19,643 @@
...
+
+import static java.nio.charset.StandardCharsets.UTF_8;
 
 
 /**
  * Datatable V4 Implementation is a wrapper around the Row-based data block.
  */
 @InterfaceStability.Evolving
-public class DataTableImplV4 extends RowDataBlock {
+public class DataTableImplV4 extends BaseDataTable {
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/datatable/DataTableImplV4.java:49]: Is V4 same as V3 + null handling ? I am guessing this is still needed even after decoupling

Comment (walterddr) [reply pinot-common/src/main/java/org/apache/pinot/common/datatable/DataTableImplV4.java:49]: Yes this class is basically a copy

**Comment by walterddr** [comment]
> > So going forward any changes on datablock will not need to consider backward compatibility against data table.
> 
> I agree that we should try to decouple. But when leaf servers do the conversion of DataTable to Block (or whatever it is evolves to in future) to send to intermediary layer of servers, this logic should still continued to be written in compatible manner. Right ?
> 
> During deployment / rolling upgrade / restart, it is possible that intermediary servers and leaf servers are having different version.

That's a good call out. Actually I have another chained PR to move the leaf servers to return non-serialized results. But this one (although refactor only) is already large enough thus I decided to split out

**Review by siddharthteotia — APPROVED** [review]

**Comment by walterddr** [comment]
merging and move on to 2nd step. which uses non-serialized results for leaf stage.

File: pinot-common/src/main/java/org/apache/pinot/common/datablock/BaseDataBlock.java [hunk 3]
```diff
@@ -442,12 +425,6 @@ public byte[] toBytes()
     DataOutputStream dataOutputStream = new DataOutputStream(byteArrayOutputStream);
     writeLeadingSections(dataOutputStream);
 
-    // Add table serialization time metadata if thread timer is enabled.
```
Comment (jasperjiaguo) [inline pinot-common/src/main/java/org/apache/pinot/common/datablock/BaseDataBlock.java:445]: Hey @walterddr  is removing this part intended?

Comment (jasperjiaguo) [inline pinot-common/src/main/java/org/apache/pinot/common/datablock/BaseDataBlock.java:445]: Hey @walterddr is removing this part intended?

Comment (walterddr) [reply pinot-common/src/main/java/org/apache/pinot/common/datablock/BaseDataBlock.java:445]: yes. 
1. we no longer encode metadata block together with data
2. data table related code are moved to DataTableImplV4

**Pull request merged**